File: haskell/src/Solutions/Euler181.hs
module Solutions.Euler181 where
import Control.Monad
import Control.Monad.ST
import Data.Array.ST

partitions bMax wMax = runST $ do
    arr <- newArray ((0, 0), (bMax, wMax)) 0 :: ST s (STArray s (Int, Int) Int)
    writeArray arr (0, 0) 1
    forM_ [0..bMax] $ \black -> forM_ [0..wMax] $ \white -> when (black + white > 0) $
        forM_ [black..bMax] $ \i -> forM_ [white..wMax] $ \j -> do
            p <- readArray arr (i - black, j - white) 
            q <- readArray arr (i, j)
            writeArray arr (i, j) (p + q)
    readArray arr (bMax, wMax)

euler181 :: IO String
euler181 = return $ show $ partitions 60 40

File: rust/src/solutions/euler181.rs
pub fn euler181() -> String {
    let b_max = 60;
    let w_max = 40;
    let mut arr = vec![vec![0u64; w_max + 1]; b_max + 1];
    arr[0][0] = 1;
    for black in 0..=b_max {
        for white in 0..=w_max {
            if black + white > 0 {
                for i in black..=b_max {
                    for j in white..=w_max {
                        arr[i][j] += arr[i - black][j - white];
                    }
                }
            }
        }
    }
    arr[b_max][w_max].to_string()
}
